### PR TITLE
fix(torch backend): reintroduced necessary manual dtype casting.

### PR DESCRIPTION
**Pull request by Madjid-CH**
Revert unnecessary changes unifyai/ivy#23659

### CODE REVIEW

File: ivy/functional/backends/torch/elementwise.py [hunk 1]
```diff
@@ -60,7 +60,8 @@ def imag(
     out: Optional[torch.Tensor] = None,
 ) -> torch.Tensor:
     if val.dtype not in (torch.complex64, torch.complex128):
-        return torch.zeros_like(val, dtype=val.dtype)
+        ret = torch.imag(val.to(torch.complex64))
```
Comment (vedpatwardhan) [inline ivy/functional/backends/torch/elementwise.py:63]: I don't think we should cast both `complex64` and `complex128` to `complex64`, we might clip large values

File: ivy/functional/backends/torch/statistical.py [hunk 2]
```diff
@@ -83,6 +83,10 @@ def mean(
             return ivy.inplace_update(out, x)
         else:
             return x
+    if "float" not in str(x.dtype) and "complex" not in str(x.dtype):  # unsupported
+        return torch.mean(x.to(torch.float32), dim=axis, keepdim=keepdims, out=out).to(
```
Comment (vedpatwardhan) [inline ivy/functional/backends/torch/statistical.py:87]: Same here, we shouldn't really hardcode type casting to `float32` on this line, it should be based on whatever's the value of `x.dtype`

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Just requested a couple of changes, thanks @Madjid-CH 😄

File: ivy/functional/backends/torch/elementwise.py (hunk 1 above)
Comment (Madjid-CH) [reply ivy/functional/backends/torch/elementwise.py:63]: yeah make sense

Comment (Madjid-CH) [reply ivy/functional/backends/torch/elementwise.py:63]: here we are casting dtypes other than `complex` to `complex64`,
I'll refactor it to make it more readble.

Comment (vedpatwardhan) [reply ivy/functional/backends/torch/elementwise.py:65]: Yeah that's an interesting point tbh, do we want to casting all other dtypes to `complex64` or just mark complex dtypes as supported for this backend if `torch.imag` doesn't support any other dtypes than `complex` ones as mentioned in their [docs](https://pytorch.org/docs/stable/generated/torch.imag.html). What do you think? Thanks @Madjid-CH 😄

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]

File: ivy/functional/backends/torch/elementwise.py (hunk 1 above)
Comment (Madjid-CH) [reply ivy/functional/backends/torch/elementwise.py:65]: if we don't want any additional memory overhead like we did with the other functions we can mark the only supported types. And I don't see any reason why this function should be an exception. So let's just mark `complex` as supported.
@MahmoudAshraf97 what do you think ?

**Review by vedpatwardhan — APPROVED** [review]
lgtm! let's see what @MahmoudAshraf97 thinks, thanks @Madjid-CH 😄

**Comment by Madjid-CH** [comment]
> Hey @Madjid-CH, could you please move this code to a branch on your fork instead of the main repo? You'd need to create a new PR from that branch to `main` instead, would be helpful to keep the main repo clean. Thanks 😄

ooh I didn't notice that, thanks for notifying me @vedpatwardhan

**Pull request closed without merging**